### PR TITLE
Add fetch_maven_artifacts plugin

### PR DESCRIPTION
**Pull request by lcarva**
Signed-off-by: Luiz Carvalho <lucarval@redhat.com>

### CODE REVIEW

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+            maven_build_path = self.path_info.mavenbuild(build_info)
+            build_archives = self.session.listArchives(buildID=build_info['id'],
+                                                       type='maven')
+            build_archives = nvr_request.match_all(build_archives)
+
+            for build_archive in build_archives:
+                maven_file_path = self.path_info.mavenfile(build_archive)
+                # NOTE: Don't use urljoin here because maven_build_path does
+                # not contain a traling slash, which causes the last dir to
```
Comment (twaugh) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: Couldn't we urljoin `maven_build_path + '/'` and `maven_file_path`? What if maven_build_path *does* contain a trailing slash in future?

(also, typo: traling)

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+                # be dropped.
+                url = maven_build_path + '/' + maven_file_path
+                download_queue.append(DownloadRequest(url, maven_file_path,
+                                                      build_archive['checksum']))
```
Comment (twaugh) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: Would there be any benefit in making this a generator function? (Although, not sure how would ever collect all errors in that case.)

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+            download_queue.append(DownloadRequest(url, target, checksum))
+
+        return download_queue
+
+    def download_files(self, downloads):
+        artifacts_path = os.path.join(self.workdir, self.DOWNLOAD_DIR)
+
+        self.log.debug('%d files to download', len(downloads))
+
+        for index, download in enumerate(downloads):
```
Comment (twaugh) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: I think the answer is no, but is there any benefit in using a thread pool here to have multiple downloads in progress at once?

File: atomic_reactor/plugins/pre_reactor_config.py [hunk 4]
```diff
@@ -128,48 +125,7 @@ def run(self):
...
-                if path.startswith('.'):
-                    path = path[1:]
-
-                self.log.error("validation error (%s): %s",
-                               path or "at top level",
-                               error.message)
-
-            raise
-
+        conf = read_yaml(config_filename, schema='schemas/config.json')
```
Comment (twaugh) [inline atomic_reactor/plugins/pre_reactor_config.py]: This could be its own commit, but if that's hard to split out it's probably fine as it is.

File: tests/plugins/test_fetch_maven_artifacts.py [hunk 5]
```diff
@@ -0,0 +1,692 @@
...
+    if not archives:
+        archives = DEFAULT_ARCHIVES
+
+    flexmock(koji, PathInfo=MockedPathInfo)
+
+    session = flexmock()
+
+    (flexmock(koji)
+        .should_receive('ClientSession')
+        # .once()
```
Comment (twaugh) [inline tests/plugins/test_fetch_maven_artifacts.py]: Why is this commented out?

File: tests/plugins/test_fetch_maven_artifacts.py [hunk 6]
```diff
@@ -0,0 +1,692 @@
...
+
+    with pytest.raises(PluginFailedException) as e:
+        runner.run()
+
+    assert 'ValidationError' in str(e)
+
+
+@responses.activate
+def test_fetch_maven_artifacts_url_with_target(tmpdir, docker_tasker):
+    """Err when downloaded remote file has unexpected checksum."""
```
Comment (twaugh) [inline tests/plugins/test_fetch_maven_artifacts.py]: Is this docstring right?

File: atomic_reactor/schemas/fetch-artifacts-nvr.json [hunk 7]
```diff
@@ -0,0 +1,56 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+
+  "title": "Artifacts to be fetched from Koji builds.",
```
Comment (twaugh) [inline atomic_reactor/schemas/fetch-artifacts-nvr.json]: We should be consistent (with other files in schemas/) in whether we use a full-stop in the title and descriptions. Currently config.json doesn't use them.

File: atomic_reactor/schemas/fetch-artifacts-nvr.json [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+  "type": "array",
+
+  "items": {
+
+    "type": "object",
+
+		"description": "A specific Koji buid and archives to be fetched.",
```
Comment (twaugh) [inline atomic_reactor/schemas/fetch-artifacts-nvr.json]: Something up with the indentation here?

Also, type: buid

File: atomic_reactor/schemas/fetch-artifacts-nvr.json [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        "type": "array",
+
+        "items": {
+
+          "type": "object",
+
+          "description": "Matcher for archives in Koji build.",
+
+          "properties": {
+
```
Comment (twaugh) [inline atomic_reactor/schemas/fetch-artifacts-nvr.json]: Why so many blank lines?

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+
+        nvr_requests = self.read_nvr_requests()
+        url_requests = self.read_url_requests()
+
+        download_queue = (self.process_by_nvr(nvr_requests) +
+                          self.process_by_url(url_requests))
+
+        self.download_files(download_queue)
+
+        # TODO: Return a list of files for koji metadata
```
Comment (twaugh) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: Yes, until the format is decided I think this is fine as-is for this PR.

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+    def process_by_url(self, url_requests):
+        download_queue = []
+
+        for url_request in url_requests:
+            url = url_request['url']
+
+            if self.allowed_domains:
+                netloc = urlparse(url).netloc.lower()
+                if netloc not in self.allowed_domains:
+                    raise ValueError('Domain {} is not in list of allowed domains: {}'
```
Comment (rcerven) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: maybe we can as well first collect all errors and raise after

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+        return util.read_yaml(file_path, default=[],
+                              schema='schemas/fetch-artifacts-url.json')
+
+    def process_by_nvr(self, nvr_requests):
+        download_queue = []
+
+        for nvr_request in nvr_requests:
+            build_info = self.session.getBuild(nvr_request.nvr)
+            if not build_info:
+                raise ValueError('Build {} not found.'.format(nvr_request.nvr))
```
Comment (vrutkovs) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: Perhaps some errors should be the showstopper - i.e its dangerous to download jars from untrusted domain

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py (hunk 3 above)
Comment (rcerven) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: well it depends from where it will be downloading, but unless each source has upload speed limit, we are limited just by our download speed, which isn't as well limited per download, so one-by one is fine without any limitations

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py (hunk 1 above)
Comment (lcarva) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: Yes, we could do that. I didn't think it was worth it. String concatenation is what koji does btw: https://pagure.io/koji/blob/master/f/cli/koji#_6721

Should I change it?

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py (hunk 2 above)
Comment (lcarva) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: Usually, a generator is used to allow processing items while the collection itself is still being created. I think in this case, we don't want that. I'd rather make sure that all the DownloadRequest instances have been created *before* actually start any downloads.

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py (hunk 3 above)
Comment (lcarva) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: It's calculating the checksum (CPU) as it's downloading (Network). A thread pool wouldn't be beneficial for splitting up checksum calculations. If going down this approach, a process pool is better suited. I'm just not sure if this is overkill at this point.

File: atomic_reactor/plugins/pre_reactor_config.py (hunk 4 above)
Comment (lcarva) [inline atomic_reactor/plugins/pre_reactor_config.py]: I'll split it up.

File: tests/plugins/test_fetch_maven_artifacts.py (hunk 5 above)
Comment (lcarva) [inline tests/plugins/test_fetch_maven_artifacts.py]: Leftover from debugging. Uncommented out.

File: tests/plugins/test_fetch_maven_artifacts.py (hunk 6 above)
Comment (lcarva) [inline tests/plugins/test_fetch_maven_artifacts.py]: Fixed it.

File: atomic_reactor/schemas/fetch-artifacts-nvr.json (hunk 7 above)
Comment (lcarva) [inline atomic_reactor/schemas/fetch-artifacts-nvr.json]: Agree. Removed punctuation.

File: atomic_reactor/schemas/fetch-artifacts-nvr.json (hunk 8 above)
Comment (lcarva) [inline atomic_reactor/schemas/fetch-artifacts-nvr.json]: Tabs! Also fixed typo.

File: atomic_reactor/schemas/fetch-artifacts-nvr.json (hunk 9 above)
Comment (lcarva) [inline atomic_reactor/schemas/fetch-artifacts-nvr.json]: My futile approach into making this file more readable. Changed it so it's more consistent with config.json.

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py (hunk 1 above)
Comment (twaugh) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: It's probably fine.

File: atomic_reactor/plugins/pre_reactor_config.py (hunk 4 above)
Comment (lcarva) [inline atomic_reactor/plugins/pre_reactor_config.py]: Done.

File: atomic_reactor/schemas/fetch-artifacts-nvr.json (hunk 9 above)
Comment (lcarva) [inline atomic_reactor/schemas/fetch-artifacts-nvr.json]: Done.

File: atomic_reactor/schemas/fetch-artifacts-nvr.json (hunk 8 above)
Comment (lcarva) [inline atomic_reactor/schemas/fetch-artifacts-nvr.json]: Done.

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py (hunk 12 above)
Comment (lcarva) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: I'll change it so it collects all errors for better user experience.

@vrutkovs I still think we should fail if the user specified a build that doesn't exist.

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+                    f.write(chunk)
+                    checksum.update(chunk)
+
+            if checksum.hexdigest() != download.md5:
+                raise ValueError('Computed md5 checksum, {}, does not match '
+                                 'expected checksum, {}'
+                                 .format(checksum.hexdigest(), download.md5))
+
+    def run(self):
+        self.session = create_koji_session(self.koji_info['hub'], self.koji_info.get('auth'))
```
Comment (twaugh) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: pylint suggests defining this in the constructor (i.e. self.session = None).

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py (hunk 12 above)
Comment (twaugh) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: We definitely should fail on error as currently. Just might be nice to report all the errors, then fail on one of them (the first?).

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py (hunk 2 above)
Comment (twaugh) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: Agreed.

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py (hunk 13 above)
Comment (lcarva) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: Done.

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py (hunk 12 above)
Comment (lcarva) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: Agreed, modified to collect all errors before raising exception.

File: atomic_reactor/plugins/pre_fetch_maven_artifacts.py (hunk 11 above)
Comment (lcarva) [inline atomic_reactor/plugins/pre_fetch_maven_artifacts.py]: Agreed, modified to collect all errors prior to raising exception.

File: atomic_reactor/util.py [hunk 14]
```diff
@@ -862,3 +866,49 @@ def are_plugins_in_order(plugins_conf, *plugins_names):
...
             return False
     return True
+
+
+def read_yaml(yaml_file_path, default=None, schema=None):
+    with open(yaml_file_path) as f:
+        data = yaml.safe_load(f)
+
+    if not schema:
+        return data
```
Comment (twaugh) [inline atomic_reactor/util.py]: Is this line ever hit?

Comment (lcarva) [inline atomic_reactor/util.py]: Hmm I guess not. I'll add a unit test for it.
I still think there's value in keep the functionality though.

Comment (lcarva) [inline atomic_reactor/util.py]: Actually, we don't have a use case for that. I'm going to make schema required to encourage always using a schema :)

Also noticed that default parameter is not used. This is leftover from a previous version. Removing it as well.

Comment (lcarva) [inline atomic_reactor/util.py]: Done!

**Comment by twaugh** [comment]
Not sure whether we want to merge this into master yet.

**Comment by lcarva** [comment]
Going over the design doc I realized the `allowed_domains` implementation was a little off. We want to be able to check the full URL prefix, not just the `netloc`. Updated accordingly, PTAL.

**Comment by lcarva** [comment]
@twaugh docs updated. Let me know if that's sufficient.
